File: reco_jet_matching/backup.py
import ROOT
import math
from TIMBER.Analyzer import analyzer
from TIMBER.Tools.Common import CompileCpp
from argparse import ArgumentParser
ROOT.gROOT.SetBatch(True)

parser=ArgumentParser()
parser.add_argument('-n', type=int, dest='n_jobs',action='store', required=True)
parser.add_argument('-i', type=int, dest='i_job',action='store', required=True)
args=parser.parse_args()
CompileCpp("jet_matching.cc")

files=[]
file_name="raw_nano/HToGG_WminusToLNu_16.txt"

with open(file_name,"r") as f:
    all_files=f.readlines()
    all_files=[line.strip() for line in all_files]
    N=len(all_files)
    print(N)
    n_apiece=math.floor(N/args.n_jobs)
    i=args.i_job
    if i<args.n_jobs-1:
        files=all_files[i*n_apiece:(i+1)*n_apiece]
    else:
        files=all_files[i*n_apiece:len(all_files)]
print(len(files))
#exit()
#files=["BF3F0C2F-BEC9-0D41-8391-1E309CBEFB4C.root"]

ana=analyzer(files)
ana.Cut("Jet","nJet>=2 || nFatJet>=1")

ana.Define("GenRelevantPartIdx","GenRelevantPart_v2(nGenPart,GenPart_pdgId,GenPart_genPartIdxMother)")
'''
#ana.Define("GenHGGJetsIdx","GenHGG(nGenPart,GenPart_pdgId,GenPart_genPartIdxMother)")
ana.Define("GenHGGJetsIdx","ROOT::VecOps::RVec<Float_t>({GenRelevantPartIdx.at(2),GenRelevantPartIdx,at(3)})")
ana.Define("GenHIdx","ROOT::VecOps::RVec<Float_t>({GenRelevantPartIdx.at(1)})")
ana.Define("GenLIdx","ROOT::VecOps::RVec<Float_t>({GenRelevantPartIdx.at(5)})")

ana.Define("Gen_JetsPt","FQuantity_match(GenHGGJetsIdx,GenPart_pt)")
ana.Define("Gen_JetsPhi","FQuantity_match(GenHGGJetsIdx,GenPart_phi)")
ana.Define("Gen_JetsEta","FQuantity_match(GenHGGJetsIdx,GenPart_eta)")
ana.Define("Gen_JetsMass","FQuantity_match(GenHGGJetsIdx,GenPart_mass)")
ana.Define("Gen_JetsInvMass","InvMass(Gen_JetsPt,Gen_JetsEta,Gen_JetsPhi,Gen_JetsMass)")

ana.Define("Gen_HPt","FQuantity_match(GenHIdx,GenPart_pt)")
ana.Define("Gen_HPhi","FQuantity_match(GenHIdx,GenPart_phi)")
ana.Define("Gen_HEta","FQuantity_match(GenHIdx,GenPart_eta)")
ana.Define("Gen_HMass","FQuantity_match(GenHIdx,GenPart_mass)")

ana.Define("Gen_LPt","FQuantity_match(GenLIdx,GenPart_pt)")
ana.Define("Gen_LPhi","FQuantity_match(GenLIdx,GenPart_phi)")
ana.Define("Gen_LEta","FQuantity_match(GenLIdx,GenPart_eta)")
ana.Define("Gen_LMass","FQuantity_match(GenLIdx,GenPart_mass)")

ana.Define("Matched_JetsIdx","Jet_match(GenHGGJetsIdx,nJet,Jet_phi,Jet_eta,GenPart_phi,GenPart_eta)")

ana.Define("Matched_JetsPt","FQuantity_match(Matched_JetsIdx,Jet_pt)")
ana.Define("Matched_JetsPhi","FQuantity_match(Matched_JetsIdx,Jet_phi)")
ana.Define("Matched_JetsEta","FQuantity_match(Matched_JetsIdx,Jet_eta)")
ana.Define("Matched_JetsMass","FQuantity_match(Matched_JetsIdx,Jet_mass)")
ana.Define("Matched_JetsInvMass","InvMass(Matched_JetsPt,Matched_JetsEta,Matched_JetsPhi,Matched_JetsMass)")

ana.Define("Matched_FatJetsIdx","FatJet_match(GenHGGJetsIdx,nFatJet,FatJet_phi,FatJet_eta,GenPart_phi,GenPart_eta)")

ana.Define("Matched_FatJetsPt","FQuantity_match(Matched_FatJetsIdx,FatJet_pt)")
ana.Define("Matched_FatJetsPhi","FQuantity_match(Matched_FatJetsIdx,FatJet_phi)")
ana.Define("Matched_FatJetsEta","FQuantity_match(Matched_FatJetsIdx,FatJet_eta)")
ana.Define("Matched_FatJetsMass","FQuantity_match(Matched_FatJetsIdx,FatJet_mass)")

ana.Define("Matched_FatJetsMassf","Matched_FatJetsMass.at(0)")
'''
'''
c_gen=ROOT.TCanvas("c","c")
c_gen.cd()
h_gen=ana.DataFrame.Filter("Gen_JetsInvMass>0").Histo1D( ('Gen_InvMass','Gen_InvMass',20,80,170), 'Gen_JetsInvMass') 
h_gen.Draw()
c_gen.Print("gen.pdf")

c_jet=ROOT.TCanvas("c","c")
c_jet.cd()
h_jet=ana.DataFrame.Filter("Matched_JetsInvMass>0").Histo1D( ('MatchedJets_InvMass','MatchedJets_InvMass',20,80,170), 'Matched_JetsInvMass') 
h_jet.Draw()
c_jet.Print("jet.pdf")

c_fatjet=ROOT.TCanvas("c","c")
c_fatjet.cd()
h_fatjet=ana.DataFrame.Filter("Matched_FatJetsMassf>0").Histo1D( ('MatchedFatJets_InvMass','MatchedFatJets_InvMass',20,80,170), 'Matched_FatJetsMassf') 
h_fatjet.Draw()
c_fatjet.Print("fatjet.pdf")

exit()
'''
#columns=["nGenPart","GenPart_eta","GenPart_phi","GenPart_pt","nJet","Jet_eta","Jet_phi","Jet_pt","nFatJet","FatJet_eta","FatJet_phi","FatJet_pt","FatJet_mass","GenHGGJetsIdx","Matched_JetsIdx","Matched_FatJetsIdx"]
#columns=["GenHGGJetsIdx","Matched_JetsIdx","Matched_FatJetsIdx"]
#columns=["nGenPart","GenPart_eta","GenPart_phi","GenPart_pt","nJet","Jet_eta","Jet_phi","Jet_pt","nFatJet","FatJet_eta","FatJet_phi","FatJet_pt","FatJet_mass","nCorrT1METJet","nSoftActivityJet","nSubJet"]
#columns=["GenHGGJetsIdx","Gen_JetsPt","Gen_JetsEta","Gen_JetsPhi","Gen_JetsMass","Gen_JetsInvMass","Matched_JetsIdx","Matched_JetsPt","Matched_JetsEta","Matched_JetsPhi","Matched_JetsMass","Matched_JetsInvMass","Matched_FatJetsIdx","Matched_FatJetsPt","Matched_FatJetsEta","Matched_FatJetsPhi","Matched_FatJetsMass"]
columns=["GenRelevantPartIdx"]
ana.Snapshot(columns,f"result_{args.n_jobs}_{args.i_job}.root","Events")

File: reco_jet_matching/backup.cc

ROOT::VecOps::RVec<Int_t> GenHGG(Int_t nGenPart,ROOT::VecOps::RVec<Int_t> GenPart_pdgId, ROOT::VecOps::RVec<Int_t> GenPart_genPartIdxMother) { \
        ROOT::VecOps::RVec<Int_t> GG={-1,-1}; \
        Int_t count=0; \
        for (Int_t i=0;i<nGenPart;i++){ \
            if (GenPart_pdgId[i]==21 && GenPart_genPartIdxMother[i]>=0 && GenPart_pdgId[GenPart_genPartIdxMother[i]]==25){ \
                GG.at(count)=i; \
                count++; \
                if(count==2) break;\
            } \
        } \
        return GG; \
} \

ROOT::VecOps::RVec<Int_t> Jet_match(ROOT::VecOps::RVec<Int_t> GenHGG,Int_t nJet,ROOT::VecOps::RVec<Float_t> Jet_phi, ROOT::VecOps::RVec<Float_t> Jet_eta,ROOT::VecOps::RVec<Float_t> GenPart_phi, ROOT::VecOps::RVec<Float_t> GenPart_eta) { \
	 Float_t R=0.4;\
        ROOT::VecOps::RVec<Int_t> GG={-1,-1}; \
        for(Int_t geni=0;geni<2;geni++){ \
	    if(GenHGG[geni]<0) continue;
            Float_t gen_phi=GenPart_phi[GenHGG[geni]]; \
            Float_t gen_eta=GenPart_eta[GenHGG[geni]];\
            for (Int_t i=0;i<nJet;i++){ \
                Float_t reco_phi=Jet_phi[i];\
                Float_t reco_eta=Jet_eta[i];\
                Float_t diff_phi=std::abs(gen_phi-reco_phi)<3.1415926 ? std::abs(gen_phi-reco_phi) : 2*3.1415926-std::abs(gen_phi-reco_phi);\
                Float_t diff_eta=std::abs(gen_eta-reco_eta);\
                if(diff_eta*diff_eta+diff_phi*diff_phi<=R*R){\
                    GG[geni]=i;\
                    break;\
                } \
            } \
        } \
        return GG; \
} \
ROOT::VecOps::RVec<Int_t> GenRelevantPart(Int_t nGenPart,ROOT::VecOps::RVec<Int_t> GenPart_pdgId, ROOT::VecOps::RVec<Int_t> GenPart_genPartIdxMother) { 
        ROOT::VecOps::RVec<Int_t> Parts={-1,-1,-1,-1,-1,-1,-1}; //Mother W, Higgs, Gluon, Gluon, Daughter W, Lepton, Neutrino 
        Int_t count=0; 
        for (Int_t i=0;i<nGenPart;i++){ 
            if (GenPart_pdgId[i]==21 && GenPart_genPartIdxMother[i]>=0 && GenPart_pdgId[GenPart_genPartIdxMother[i]]==25){ 
                Parts.at(count+2)=i;
	        Parts.at(1)=GenPart_genPartIdxMother[i];	
                count++; 
                if(count==2) break;
            } 
        }
       	if(Parts.at(1)>0){
			Parts.at(0)=GenPart_genPartIdxMother[Parts.at(1)];
			if(Parts.at(0)>0){
				for (Int_t i=0;i<nGenPart;i++){
					if (GenPart_pdgId[i]==-24 && GenPart_genPartIdxMother[i]==Parts.at(0)){
						Parts.at(4)=i;
					 	for (Int_t j=0;j<nGenPart;j++){
							if (GenPart_genPartIdxMother[j]==Parts.at(4)){
								 if(GenPart_pdgId[j]==11 || GenPart_pdgId[j]==13 || GenPart_pdgId[j]==15)
									 Parts.at(5)=j;
							 if(GenPart_pdgId[j]==-12 || GenPart_pdgId[j]==-14 || GenPart_pdgId[j]==-16)
								 Parts.at(6)=j;
						 }
					 }
					 break;
				}
			}
		}

	}	
        return Parts; 
} 
Int_t FatJetMatching_deltaR(RVec<Int_t> GenHGG, RVec<Float_t> GenPart_phi, RVec<Float_t> GenPart_eta, Int_t nFatJet, RVec<Float_t> FatJet_phi, RVec<Float_t> FatJet_eta,  Float_t R=0.8) { 
        ROOT::VecOps::RVec<Int_t> GG={-1}; 
	if(GenHGG[0]<0 || GenHGG[1]<0) return GG;
        for (Int_t i=0;i<nJet;i++){ 
	    int matched=1;
            for(Int_t geni=0;geni<2;geni++){ 
            	Float_t gen_phi=GenPart_phi[GenHGG[geni]]; 
            	Float_t gen_eta=GenPart_eta[GenHGG[geni]];
                Float_t reco_phi=Jet_phi[i];
                Float_t reco_eta=Jet_eta[i];
                Float_t diff_phi=std::abs(gen_phi-reco_phi)<3.1415926 ? std::abs(gen_phi-reco_phi) : 2*3.1415926-std::abs(gen_phi-reco_phi);
                Float_t diff_eta=std::abs(gen_eta-reco_eta);
                if(diff_eta*diff_eta+diff_phi*diff_phi>R*R)
                    matched=0; 
            } 
	    if (matched==1){
		    GG.at(0)=i;
		    break;
	    }
        } 
        return GG; 
} 
